### ISSUE TITLE
vae loss increases

### ISSUE THREAD

**Issue by Bazinga-ai** [EdoardoBotta/RQ-VAE-Recommender#30]
Hi Botta, Thank you for your work. I encountered an issue during training. I noticed that the overall loss gradually decreases and stabilizes after around 1 epoch. However, the validation loss (VL loss) initially decreases but then starts increasing before stabilizing again after 1 epoch. Do you know why this happens?

**Comment by EdoardoBotta** [comment]
Hi @Bazinga-ai , thank you for reaching out. Based on the frequency of these plots, I'm assuming these are training loss curves but correct me if I'm wrong and you're doing eval more frequently. Anyways, similar behavior is expected both at training and eval time.

In the code, vl does not stand for Validation Loss, but for VAE Loss (defined [here](https://github.com/EdoardoBotta/RQ-VAE-Recommender/blob/main/modules/loss.py#L34)). It is the loss components that ensures the encoder and codebook vectors are close to the each other.   

On the other hand, rl stands for Reconstruction Loss and we have (loss = vl + rl). There are two parts to the question as to why VAE loss starts low and increases over time:
* **Why does it start low?** This is due to codebook initialization. The codebook is initialized by performing a KMeans clustering. If you take a look at the definition of the VAE loss, for fixed encoder output, this is exactly the loss that KMeans minimizes. Hence, it's not surprising that the initial value for vl is relatively low.
* **Why does it increase over time?** This is because the model is optimizing the sum of two competing loss terms, vl and rl. It is possible, and indeed this what's happening, that the optimization algorithm moves to parameters with slightly increased VAE loss but drastically lower reconstruction loss than the initial KMeans. This is also encouraged by the scale of the two losses. A way you could prioritize VAE loss over Reconstruction loss is by increasing the weight of the VAE loss. As you increase such weight, you would see the tradeoff between the two losses move in favor of VAE loss.

Hope it helps.
